Added Dot Layer

therealansh opened the pull request:
This PR adds Dot layer and closes #144. @zaleslaw would you mind reviewing it for this initial commit. I am still working on some test cases.

zaleslaw commented on api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/merge/Dot.kt:
```diff
@@ -0,0 +1,126 @@
+import org.tensorflow.op.math.Rsqrt
+import org.tensorflow.op.math.Square
+
+public class Dot(
```
Please add KDocs (look at Keras documenttion for consistency)

zaleslaw commented on api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/merge/Dot.kt:
```diff
@@ -0,0 +1,126 @@
+
+public class Dot(
+    val axis: IntArray,
+    val normalize: Boolean = false,
```
make this field public because it's a part of public API and add KDocs

zaleslaw commented on api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/merge/Dot.kt:
```diff
@@ -0,0 +1,126 @@
+
+    constructor(axis: Int) : this(axis = IntArray(2) { axis })
+
+    private fun l2Normalize(scope: Scope?, x: Operand<Float>, axis: IntArray?): Operand<Float> {
```
Add documentation KDocs for this function, move out of class and make public (it could be reused as a building block)

therealansh replied:
`l2normalize` and `batchdot` are present in linalg and backend in respectively in the official Keras.Where should i refactor them out of here? Or should i just refactor them out of this class ad put it in `Dot.kt` only?

zaleslaw replied:
I agree that there is no best place for them at this moment (and other missed, implemented-in-Keras only parts). Could you please just refactor them out of class and put it in Dot.kt like you suggested

zaleslaw commented on api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/merge/Dot.kt:
```diff
@@ -0,0 +1,126 @@
+        return Mul.create(scope, x, invNorm)
+    }
+
+    private fun batchDot(scope: Scope?, x: Operand<Float>, y: Operand<Float>, axis: IntArray): Operand<Float> {
```
Probably this function could be moved out of class, made public and KDocs could be useful too

zaleslaw commented:
Hi @therealansh do you have plans to fix a review comments here? It will be great to merge it in December in the codebase

therealansh commented:
hey @zaleslaw sorry about this delay, i am actually figuring a way out for test cases I'll fix this and add docs accordingly. Maybe i can add `DotTestCase` in PR for #276 what do you think?

therealansh commented:
Added the docs! I have removed the Dot Test case for now. I'll be doing it in the PR for #276. Should i add a TODO as well for refactoing l2Normalize and batchDot for future refactoring?

zaleslaw commented:
No needs to add TODO, I make this refactoring in the nearest future

zaleslaw commented:
Great, now we have DOT!

therealansh commented:
Hey @zaleslaw , while writing test cases for `Dot` i have run into some errors,
```
In[0] is not a matrix. Instead it has shape [2,1,5,2]
```
at
```kt
 out = MatMul.create(scope, x2, y2, MatMul.transposeA(adjX), MatMul.transposeB(adjY))
 ```
 in batchDot. Can you review it once I cannot understand what the issue is about?
